View breadth requirements for UTSC courses

chintans1 opened the pull request:
This change will allow for courses that are at UTSC to have breadth requirements. Previously the scraper would only fetch breadths for UTSG courses. I have changed that now so we can populate breadths for UTSC courses as well. It still holds the model validations so "breadths" is still a list of numbers. 

For UTSC, the numbers correspond the breadth in this fashion:
1 => Arts, Literature & Language
2 => "History, Philosophy & Cultural Studies"
3 => "Natural Sciences"
4 => "Social & Behavioural Sciences"
5 => "Quantitative Reasoning"

UTSC courses don't have more than one breadth so I feel this change should work for all courses at UTSC. I have tested it and confirmed the outputted data to be correct.

chintans1 committed:
View breadth requirements for UTSC courses

This change will allow for courses that are at UTSC to have breadth requirements. Previously the scraper would only fetch breadths for UTSG courses. I have changed that now so we can populate breadths for UTSC courses as well. It still holds the model validations so "breadths" is still a list of numbers. 

For UTSC, the numbers correspond the breadth in this fashion:
1 => Arts, Literature & Language
2 => "History, Philosophy & Cultural Studies"
3 => "Natural Sciences"
4 => "Social & Behavioural Sciences"
5 => "Quantitative Reasoning"

UTSC courses don't have more than one breadth so I feel this change should work for all courses at UTSC. I have tested it and confirmed the outputted data to be correct.

g3wanghc commented:
@chintans1 I think UTM also has a [Distribution Requirements](https://www.utm.utoronto.ca/registrar/academic-support/distribution-requirements), it might be worthwhile to just handle all cases in tri-campus

chintans1 commented:
Yea, but I feel that is would be wrong to have UTM Distribution Requirement under breadths. The list of numbers would not directly correlate to the distribution. For UTSC, the breadths are basically the same with the just the wording different but UTM sort of has a different system altogether.

g3wanghc commented:
@chintans1 maybe you can handle this by using different fields to avoid any ambiguity

chintans1 commented:
@g3wanghc I suppose, but also, the UTM Distribution field does not appear for all courses at UTM and I don't see another field to use for this unless I add one myself which would I think require work in the scraper and the API itself.

arkon commented:
Yes, that's beyond the scope of this PR.

arkon commented on uoftscrapers/scrapers/courses/__init__.py:
```diff
@@ -124,16 +124,33 @@ def parse_course_html(course_id, html):
+            utsc_breadth = utsc_breadth.find_all("span", id="u104")[0] \
+                .get_text().strip()
+
+            if utsc_breadth == "Arts, Literature & Language":
```
It'd be nice if this was in some sort of enum mapping.

chintans1 committed:
Clean-up the if-statement for finding UTSC breadth

Instead of having one big if statement to cover all the UTSC breadth categories, I decided to use a dictionary to store the category names and assign it to a value. This makes the code much cleaner and we can easily add the breadth number to the list.

chintans1 commented:
@arkon I've made some changes to clean it up. I think it is better now with the usage of a dictionary

kashav commented:
~Any chance that a course will have more than one breadth category?~

Edit: Oops, just read the PR description 😅, LGTM.

chintans1 commented:
@kshvmdn I have yet to see any UTSC course cover more than one breadth which made me do it this way.

kashav commented:
@chintans1 I think I can confirm that -- all of the following appear to be empty.

- [`BRE_ID_A1153,BRE_ID_H1169`](http://coursefinder.utoronto.ca/course-search/search/courseSearch/course/search?queryText=&requirements=BRE_ID_A1153,BRE_ID_H1169&campusParam=Scarborough)
- [`BRE_ID_A1153,BRE_ID_N1109`](http://coursefinder.utoronto.ca/course-search/search/courseSearch/course/search?queryText=&requirements=BRE_ID_A1153,BRE_ID_N1109&campusParam=Scarborough)
- [`BRE_ID_A1153,BRE_ID_Q1118`](http://coursefinder.utoronto.ca/course-search/search/courseSearch/course/search?queryText=&requirements=BRE_ID_A1153,BRE_ID_Q1118&campusParam=Scarborough)
- [`BRE_ID_A1153,BRE_ID_S1124`](http://coursefinder.utoronto.ca/course-search/search/courseSearch/course/search?queryText=&requirements=BRE_ID_A1153,BRE_ID_S1124&campusParam=Scarborough)
- [`BRE_ID_H1169,BRE_ID_N1109`](http://coursefinder.utoronto.ca/course-search/search/courseSearch/course/search?queryText=&requirements=BRE_ID_H1169,BRE_ID_N1109&campusParam=Scarborough)
- [`BRE_ID_H1169,BRE_ID_Q1118`](http://coursefinder.utoronto.ca/course-search/search/courseSearch/course/search?queryText=&requirements=BRE_ID_H1169,BRE_ID_Q1118&campusParam=Scarborough)
- [`BRE_ID_H1169,BRE_ID_S1124`](http://coursefinder.utoronto.ca/course-search/search/courseSearch/course/search?queryText=&requirements=BRE_ID_H1169,BRE_ID_S1124&campusParam=Scarborough)
- [`BRE_ID_N1109,BRE_ID_Q1118`](http://coursefinder.utoronto.ca/course-search/search/courseSearch/course/search?queryText=&requirements=BRE_ID_N1109,BRE_ID_Q1118&campusParam=Scarborough)
- [`BRE_ID_N1109,BRE_ID_S1124`](http://coursefinder.utoronto.ca/course-search/search/courseSearch/course/search?queryText=&requirements=BRE_ID_N1109,BRE_ID_S1124&campusParam=Scarborough)
- [`BRE_ID_Q1118,BRE_ID_S1124`](http://coursefinder.utoronto.ca/course-search/search/courseSearch/course/search?queryText=&requirements=BRE_ID_Q1118,BRE_ID_S1124&campusParam=Scarborough)

chintans1 commented:
That's great. Hopefully this can be merged then! :)

arkon commented on uoftscrapers/scrapers/courses/__init__.py:
```diff
@@ -124,15 +124,32 @@ def parse_course_html(course_id, html):
                 if ch in "12345":
                     breadths.append(int(ch))
 
+        # If Campus is UTSC, find Breadths
```
Might want to put this into an `elif` branch since it wouldn't make sense to run this at all if we already did UTSG above.

arkon reviewed (changes requested):
Just one small change and we should be good to go.

chintans1 committed:
Clean up the if-statement for finding breadths

 - make sure its run once and we do not declare anything that is not needed

Signed-off-by: Chintan Shah <chintans98@gmail.com>

chintans1 commented:
@arkon I think that is better. I made it so we don't declare any variables that we won't use. We find the breadths now for only UTSG and UTSC courses so UTM is always empty.

qasim commented:
LGTM.

Thanks for the contribution @chintans1 :) always appreciate help to improve UTSC/UTM data.
